### PR TITLE
[test] add cross-ref tests for python meta kernels

### PR DESCRIPTION
**Pull request by bdhirsh**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #86611
* #86610
* #86609
* __->__ #86228
* #86608
* #86607
* #86285

### CODE REVIEW

File: torch/_refs/__init__.py [hunk 1]
```diff
@@ -906,13 +906,14 @@ def add(
...
         dtype = a.dtype if isinstance(a, TensorLike) else b.dtype  # type: ignore[union-attr]
         python_type = utils.dtype_to_type(dtype)
-        if not utils.is_weakly_lesser_type(type(alpha), python_type):
-            msg = (
-                "alpha argument of type {0} cannot be safely cast to type {1}!".format(
-                    type(alpha), python_type
-                )
-            )
-            raise ValueError(msg)
+        # TODO: this check is too restrictive. torch.add(bool_tensor, bool_tensor, 0) should work.
```
Comment (bdhirsh) [inline torch/_refs/__init__.py:909]: cc @ngimel. This check triggers on `torch.add(bool_tensor, bool_tensor, 0)`, which works fine on eager. My understanding is that refs should work in all cases that eager works - do you happen to know the purpose of this check?

File: torch/_decomp/decompositions.py [hunk 2]
```diff
@@ -1776,10 +1780,11 @@ def index_add_(
     alpha: NumberType = 1,
 ):
     dim = utils.canonicalize_dims(x.ndim, dim)
-    utils.check(
-        index.ndim <= 1,
-        lambda: f"Index should have dimension 1 or 0 (got {index.ndim})",
-    )
+    # Fails incorrectly for bool tensor
```
Comment (albanD) [inline torch/_decomp/decompositions.py:1783]: What do you mean?

Comment (bdhirsh) [reply torch/_decomp/decompositions.py:1779]: @albanD This is the same issue as commented out check on like 909. `addmm(bool_tensor1, bool_tensor2, bool_tensor3, beta=1, alpha=1)` fails this check, but works fine in eager mode.

Do you know why this check was added? Are there some checks in C++ that we're trying (incorrectly) to emulate?

Comment (albanD) [reply torch/_decomp/decompositions.py:1779]: I am not sure tbh. But I would expect that yes. I'm sure Natalia would know.

**Review by albanD — APPROVED** [review]
SGTM but should get a second opinion from Natalia

File: torch/_refs/__init__.py (hunk 1 above)
Comment (mruberry) [reply torch/_refs/__init__.py:909]: We need to modify these checks to account for PyTorch allowing integer numbers with boolean tensors. Typically a number cannot be a "higher" Python type than the Python type corresponding to the tensors' type. For example,

```
a = torch.randint(0, 5, (2, 2))
b = torch.randint(0, 5, (2, 2))
torch.add(a, b, alpha=.5)
: RuntimeError: For integral input tensors, argument alpha must not be a floating point number.
```

The exception to this is that boolean tensors will interoperate with integers, and I think the integer is always cast to a bool before the operation. The references just don't implement this particular oddity yet.

File: torch/_decomp/decompositions.py (hunk 2 above)
Comment (ngimel) [reply torch/_decomp/decompositions.py:1779]: Why do we even support addmm for bools? I'm pretty sure we don't support it on cuda.

Comment (lezcano) [reply torch/_decomp/decompositions.py:1779]: This is the check, I believe:
https://github.com/pytorch/pytorch/blob/f903f1ab343fea72177f29fc8d453febcaad8905/aten/src/ATen/native/TensorAdvancedIndexing.cpp#L284-L285
As to why does it fails for bools... I don't know.

I am currently putting together a more thorough test for decompositions. This test that has uncovered a few issues in some operations, in particular for bools. No idea what's going on there just yet. I plan to fix all the issues the test has uncovered before putting up a PR. Hopefully that'll be today or tomorrow. For now, I'm leaving here this fantastic behaviour it uncovered:

```python
import torch msk = torch.tensor([False])
b = torch.tensor([False])
print(torch.where(msk, True, b).dtype)
print(torch.ops.aten.where.ScalarSelf(msk, True, b).dtype)
print(torch._refs.where(msk, True, b).dtype)
```
outputs
```
torch.bool
torch.int64
torch.bool
```
This happens just for boolean tensors, all the other types work just fine.

File: torch/_refs/__init__.py (hunk 1 above)
Comment (bdhirsh) [reply torch/_refs/__init__.py:909]: thanks for the overview! I'll update this check to pass in the bool case.

**Pull request closed without merging**